File: python/game-of-life/game_of_life.py
"""
Game of Life
For a given 2D matrix return the next state in the game of life simulation.
"""

def tick(matrix):
    """Run the game of life simulation one step returning a matrix with next next output state.
    :param matrix: Current state of the game of life simulation
    :returns: Next state of the game of life simulation as a 2D array of 0, 1 values 1 - alive 0 - dead.
    """
    next_matrix = []
    if len(matrix) > 0:
        next_matrix = new_matrix(len(matrix), max(len(row) for row in matrix), 0)
    for row_index, row in enumerate(matrix):
        for column_index, col in enumerate(row):
            neighbor_count = count_neighbors(matrix, row_index, column_index)
            next_matrix[row_index][column_index] = int((col != 0 and neighbor_count == 2) or neighbor_count == 3)
    return next_matrix

def new_matrix(rows, cols, fill):
    """
    Create a new 2D matrix of size rows x cols where all the
    cells are set to fill.
    :param rows: How many rows the new matrix should have.
    :param cols: How many columns the new matrix should have.
    :param fill: What to fill the matrix with.
    :returns: A new 2D matrix filled with default value
    """
    return [[fill for column in range(cols)] for row in range(rows)]

def get_matrix(matrix, row, column, default):
    """
    Get a cell from a 2D matrix. Returning default if 
    the row/column is out of bounds.
    :param matrix: The matrix to get a value from
    :param row: The row index to retrieve from
    :param column: The column index to retrieve from
    :param default: Value to return if out of bounds
    :returns: Matrix[row, column] or default if out of bounds
    """
    if row < 0 or column < 0 or row >= len(matrix) or column >= len(matrix[row]):
        return default
    return matrix[row][column]

def count_neighbors(matrix, row, column):
    """
    Get the number of living neighbors for a position in the matrix
    :param matrix: The matrix to search (2-D list)
    :param row: row index to search from
    :param column: column index to search from
    :returns: The number of living neighbors of the given cell (itself not included)
    """
    return sum(get_matrix(matrix, row_index, column_index, 0) for row_index in range(row - 1, row + 2) for column_index in range(column - 1, column + 2) if row_index != row or column_index != column)

File: go/game-of-life/game_of_life.go
// Game of Life exercise
package gameoflife

// Calculate the next state of the game board in the game of life.
// @param matrix: The current state of the game as a nxm slice of integer slices where
// 1 = alive and 0 = dead.
// @returns: new nxm slice of integer slices filled with the calculated state of the next
// step of the simulation.
func Tick(matrix [][]int) [][]int {
	var rows = len(matrix)
	var ret = make([][]int, rows)
	for j := range rows {
		ret[j] = make([]int, len(matrix[j]))
		for i := 0; i < len(matrix[j]); i++ {
			ret[j][i] = Next(matrix, i, j)
		}
	}

	return ret
}

// Calcuate the next state of a cell in the game of life.
// @param matrix: The current game of life board
// @param x: x coordinate of the position to evaluate
// @param y: y coordinate of the position to evaluate
// @returns: Integer 1 if alive or 0 if dead on the next iteration.
func Next(matrix [][]int, x int, y int) int {
	var countAlive int = 0
	var isAlive = matrix[y][x]
	for j := y - 1; j <= y+1; j++ {
		if j < 0 || j >= len(matrix) {
			continue
		}
		for i := x - 1; i <= x+1; i++ {
			if i < 0 || i >= len(matrix[j]) {
				continue
			}
			if i == x && j == y {
				continue
			}
			countAlive += matrix[j][i]
		}
	}
	if isAlive == 1 && (countAlive == 2 || countAlive == 3) {
		return 1
	}
	if isAlive == 0 && countAlive == 3 {
		return 1
	}
	return 0
}
